src/profile-logic/profile-data.js

```diff
@@ -169,297 +168,404 @@ type CallNodeTableAndStackMap = {
       }
-      stackIndexToCallNodeIndex[stackIndex] = callNodeIndex;
+      currentLastChild[prefixCallNode] = callNodeIndex;
     }
-    return _createCallNodeTableFromUnorderedComponents(
-      prefix,
-      firstChild,
-      nextSibling,
-      func,
-      category,
-      subcategory,
-      innerWindowID,
-      sourceFramesInlinedIntoSymbol,
-      length,
-      stackIndexToCallNodeIndex
-    );
-  });
+  }
+  return {
+    prefix,
+    firstChild,
+    nextSibling,
+    length,
+    stackIndexToCallNodeIndex,
+  };
 }
 
 /**
- * Create a CallNodeTableAndStackMap with an ordered call node table based on
- * the pieces of an unordered call node table.
+ * Used as part of creating the call node table. This function computes the
+ * final order of call nodes, and returns columns which describe the tree
+ * structure with that final order, i.e. in DFS order. DFS here means
+ * "depth-first search":
+ *
+ *  - If a node A has children, its first child B directly follows A.
+ *  - Otherwise, the node following A is A's next sibling (if it has one), or
+ *    the next sibling of the closest ancestor which has a next sibling.
  *
- * The order of siblings is maintained.
- * If a node A has children, its first child B directly follows A.
- * Otherwise, the node following A is A's next sibling (if it has one), or the
- * next sibling of the closest ancestor which has a next sibling.
- * This means that any node and all its descendants are laid out contiguously.
+ * This means that for any node, the node and all its descendants are laid out
+ * contiguously. This contiguous chunk is described by the `subtreeRangeEnd`
+ * column and allows other parts of the codebase to perform cheap "is descendant"
+ * checks.
+ *
+ * We do not order siblings by func. The order of siblings is meaningless, and
+ * is based on the somewhat arbitrary order in which we encounter the original
+ * stack nodes in the stack table.
  */
-function _createCallNodeTableFromUnorderedComponents(
-  prefix: Array<IndexIntoCallNodeTable>,
-  firstChild: Array<IndexIntoFuncTable>,
-  nextSibling: Array<IndexIntoFuncTable>,
-  func: Array<IndexIntoFuncTable>,
-  category: Array<IndexIntoCategoryList>,
-  subcategory: Array<IndexIntoSubcategoryListForCategory>,
-  innerWindowID: Array<InnerWindowID>,
-  sourceFramesInlinedIntoSymbol: Array<IndexIntoNativeSymbolTable | -1 | -2>,
-  length: number,
-  stackIndexToCallNodeIndex: Int32Array
-): CallNodeTableAndStackMap {
-  return timeCode('createCallNodeInfoFromUnorderedComponents', () => {
-    if (length === 0) {
-      return {
-        callNodeTable: getEmptyCallNodeTable(),
-        stackIndexToCallNodeIndex: new Int32Array(0),
-      };
+function _computeCallNodeTableDFSOrder(
+  hierarchy: CallNodeTableHierarchy
+): CallNodeTableDFSOrder {
+  const { prefix, firstChild, nextSibling, length, stackIndexToCallNodeIndex } =
+    hierarchy;
+
+  if (length === 0) {
+    throw new Error('Empty call node table hierarchy');
```

Comment:
I'm a bit worried by this throw. Would it be possible to return something sensible instead?

Reply:
Ok, replaced this with a return.